core/chaincode/lifecycle/lifecycle.go

```diff
@@ -496,6 +504,90 @@ func (ef *ExternalFunctions) ApproveChaincodeDefinitionForOrg(chname, ccname str
 	return nil
 }
 
+// QueryApprovedChaincodeDefinition returns the approved chaincode definition in Org state by using the given parameters.
+// If the parameter of sequence is not provided, this function returns the latest approved chaincode definition
+// (latest: new one of the currently defined sequence number and the next sequence number).
+func (ef *ExternalFunctions) QueryApprovedChaincodeDefinition(chname, ccname string, sequence int64, publicState ReadableState, orgState ReadableState) (*ApprovedChaincodeDefinition, error) {
+
+	requestedSequence := sequence
+
+	// If requested sequence is not provided,
+	// set the latest sequence number (either the currently defined sequence number or the next sequence number)
+	if requestedSequence == 0 {
+		currentSequence, err := ef.Resources.Serializer.DeserializeFieldAsInt64(NamespacesName, ccname, "Sequence", publicState)
+		if err != nil {
+			return nil, errors.WithMessage(err, "could not get current sequence")
+		}
+		requestedSequence = currentSequence
+
+		nextSequence := currentSequence + 1
+		privateName := fmt.Sprintf("%s#%d", ccname, nextSequence)
+		_, ok, err := ef.Resources.Serializer.DeserializeMetadata(NamespacesName, privateName, orgState)
+		if err != nil {
+			return nil, errors.WithMessagef(err, "could not deserialize namespace metadata for next sequence %d", nextSequence)
+		}
+		if ok {
+			requestedSequence = nextSequence
+		}
+	}
+
+	logger.Infof("Attempting to fetch approved definition (name: '%s', sequence: '%d') on channel '%s'", ccname, requestedSequence, chname)
+	privateName := fmt.Sprintf("%s#%d", ccname, requestedSequence)
+	metadata, ok, err := ef.Resources.Serializer.DeserializeMetadata(NamespacesName, privateName, orgState)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "could not deserialize namespace metadata for %s", privateName)
+	}
+
+	if ok {
+		// Get chaincode parameters for the request sequence
+		ccParameters := &ChaincodeParameters{}
+		if err := ef.Resources.Serializer.Deserialize(NamespacesName, privateName, metadata, ccParameters, orgState); err != nil {
```

Comment:
Although it should not currently be possible, we need to check the metadata type to ensure that this approval corresponds to a chaincode definition before attempting to deserialize.  Something like:

```
        if metadata.Datatype != ChaincodeDefinitionType {
                return false, nil, errors.Errorf("not a chaincode type: %s", metadata.Datatype)
        }
```

Reply:
The new patch added some logic checking the metadata type.
Instead of ChaincodeDefinition, ApproveChaincodeDefinitionForOrg stores ChaincodeParameters and ChaincodeLocalPackage as the part of the chaincode definition in the Org state. So, the added logic checks the above types.